smart-contract-verifier/src/solidity/solc_cli.rs

```diff
@@ -0,0 +1,493 @@
+
+use std::{
+    collections::BTreeMap,
+    path::PathBuf,
+    process::{Command, Stdio},
+};
+
+use ethers_solc::{
+    artifacts::Severity,
+    error::{SolcError, SolcIoError},
+    CompilerInput, CompilerOutput,
+};
+
+mod serde_helpers {
+    use serde::de;
+
+    pub fn deserialize_abi_string<'de, D>(
+        deserializer: D,
+    ) -> Result<Vec<serde_json::Value>, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let s: String = de::Deserialize::deserialize(deserializer)?;
+        serde_json::from_str(&s).map_err(de::Error::custom)
+    }
+}
+
+mod types {
+    use super::serde_helpers;
+    use std::collections::{BTreeMap, HashMap};
+
+    use ethers_solc::{
+        artifacts::{Contract, Libraries},
+        error::SolcError,
+        CompilerInput, CompilerOutput,
+    };
+    use minicbor::encode::Write;
+    use serde::{Deserialize, Serialize};
+    use tempfile::TempDir;
+
+    #[derive(Debug, PartialEq, Eq)]
+    pub struct InputArgs {
+        pub optimize: bool,
+        pub optimize_runs: Option<usize>,
+        pub libs: BTreeMap<String, String>,
+    }
+
+    fn merge_libs(libraries: Libraries) -> BTreeMap<String, String> {
+        libraries
+            .libs
+            .into_iter()
+            .map(|(_, libs)| libs)
+            .reduce(|accum, item| accum.into_iter().chain(item).collect())
+            .unwrap_or_default()
+    }
+
+    impl TryFrom<&CompilerInput> for InputArgs {
+        type Error = SolcError;
+        fn try_from(input: &CompilerInput) -> Result<Self, Self::Error> {
+            let libs = merge_libs(input.settings.libraries.clone());
+            Ok(InputArgs {
+                optimize: input.settings.optimizer.enabled.unwrap_or(false),
+                optimize_runs: input.settings.optimizer.runs,
+                libs,
+            })
+        }
+    }
+
+    impl InputArgs {
+        pub fn build_string_args(&self) -> Vec<String> {
+            let mut vec: Vec<String> = vec![
+                "--combined-json".to_string(),
+                OutputContract::keys().join(","),
+            ];
+            if self.optimize {
+                vec.push("--optimize".to_string());
+            }
+            if let Some(runs) = self.optimize_runs {
+                // Note: --optimize-runs doesn't affect bytecode without --optimize
+                vec.push("--optimize-runs".to_string());
+                vec.push(runs.to_string());
+            }
+            if !self.libs.is_empty() {
+                vec.push("--libraries".to_string());
+                let libs: Vec<String> = self
+                    .libs
+                    .iter()
+                    .map(|(name, address)| format!("{}:{}", name, address))
+                    .collect();
+                vec.push(libs.join(","))
+            }
+            vec
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct InputFiles {
+        pub files_dir: TempDir,
+        pub file_names: Vec<String>,
```

Comment:
Why not to make it `PathBuf`?
```suggestion
        pub file_names: Vec<PathBuf>,
```

Reply:
eventually in `build_string_args` I have to convert it to string anyway

Reply:
You can make `build_string_args` to return `Vec<PathBuf>` without any problem